azafea/event_processors/endless/metrics/v3/model/_base.py

```diff
@@ -191,7 +191,7 @@ class MetricEvent(Base, metaclass=MetricMeta):
     __abstract__ = True
 
     __event_uuid__: str
-    __payload_type__: Optional[str]
+    __payload_type__: Optional[Union[str, Tuple[str, ...]]]
```

Comment:
This seems useful, but my interpretation is that you can't really have an event processor handle arbitrary payloads. It seems like the limitation here is that all the payload types must be able to be mapped to the same database schema.

In the past the way we would have handled this would be a `FooEvent2` with the new payload type, right?

Anyways, I think this makes sense. I can see how having the event processor handle multiple event UUIDs would be better in some ways, but I can see the utility of both should someone want to do the larger refactoring at some point. It's a lot of churn to add a new UUID when the processor can just handle different payloads appropriately. Since you're still type checking the payload regardless of how many types there may be, it seems to me that the multiple UUIDs per processor really only benefits the analysis side where you can munge multiple events into the same table. Is my understanding of this correct?

Reply:
> It seems like the limitation here is that all the payload types must be able to be mapped to the same database schema.

Yep.

> In the past the way we would have handled this would be a `FooEvent2` with the new payload type, right?

Yes. Or: change the UUID, migrate the existing data, and discard any future submissions of data with the old UUID.

In this case it seems unnecessary to do any migration when the database schema already supports the change we want to make; and it's easy enough to keep receiving new data in the old format.

> Anyways, I think this makes sense. I can see how having the event processor handle multiple event UUIDs would be better in some ways, but I can see the utility of both should someone want to do the larger refactoring at some point. It's a lot of churn to add a new UUID when the processor can just handle different payloads appropriately. Since you're still type checking the payload regardless of how many types there may be, it seems to me that the multiple UUIDs per processor really only benefits the analysis side where you can munge multiple events into the same table. Is my understanding of this correct?

At present you have (in theory) the nice property that each event UUID always has a payload of a single known type, and if the payload type doesn't match you can probably just discard the event. The payloads are transmitted within a value of type `mv`, so you can always deserialize them with or without this property.

I say "in theory" because I have a dim memory that there were times in the past where we changed the type of an event's payload without changing its UUID. In one case I guess we changed an event from "no payload" to "has payload", which is why we have special handling of the `EmptyPayloadError` where if the UUID is in a `IGNORED_EMPTY_PAYLOAD_ERRORS` array we just discard the event.

Right now we (otherwise) don't discard events whose payload is of an unexpected type: they get stashed in the invalid_[singular|aggregate]_events_v3 tables, and we replay them every time we redeploy azafea. (Actually I just noticed that we never implemented that replay logic for aggregate events, whoops.)